decomfl_main.py

```diff
@@ -9,31 +9,57 @@
 from byzantine import aggregation as byz_agg
 from byzantine import attack as byz_attack
 from cezo_fl.client import ResetClient
-from cezo_fl.fl_helpers import get_client_name
+from cezo_fl.fl_helpers import get_client_name, get_server_name
 from cezo_fl.server import CeZO_Server
-from cezo_fl.util import model_helpers, prepare_settings
-
-from config import get_args_str, get_params
-from preprocess import preprocess
+from cezo_fl.util import model_helpers
+from experiment_helper import prepare_settings
+from experiment_helper.cli_parser import (
+    GeneralSetting,
+    DeviceSetting,
+    DataSetting,
+    ModelSetting,
+    OptimizerSetting,
+    FederatedLearningSetting,
+    RGESetting,
+    ByzantineSetting,
+)
+from experiment_helper.device import use_device
+from experiment_helper.data import get_dataloaders
+
+
+class CliSetting(
+    GeneralSetting,
+    DeviceSetting,
+    DataSetting,
+    ModelSetting,
+    OptimizerSetting,
+    FederatedLearningSetting,
+    RGESetting,
+    ByzantineSetting,
+):
+    pass
 
 
 def setup_server_and_clients(
-    args, device_map: dict[str, torch.device], train_loaders
+    args: CliSetting, device_map: dict[str, torch.device], train_loaders
 ) -> CeZO_Server:
     model_inferences, metrics = prepare_settings.get_model_inferences_and_metrics(
-        args.dataset, prepare_settings.SUPPORTED_LLM.get(args.large_model)
+        args.dataset, args
     )
     clients = []
 
     for i in range(args.num_clients):
         client_name = get_client_name(i)
         client_device = device_map[client_name]
-        (
-            client_model,
-            client_optimizer,
-            client_grad_estimator,
-        ) = prepare_settings.prepare_settings_underseed(args, client_device)
-        client_model.to(client_device)
+        client_model = prepare_settings.get_model(
+            dataset=args.dataset, model_setting=args, seed=args.seed
+        ).to(client_device)
+        client_optimizer = prepare_settings.get_optimizer(
+            model=client_model, dataset=args.dataset, optimizer_setting=args
+        )
+        client_grad_estimator = prepare_settings.get_random_gradient_estimator(
+            model=client_model, device=client_device, rge_setting=args, model_setting=args
+        )
```

Comment:
I like this style but the drawback is this `model_setting=args`, `optimizer_setting=args`, `rge_setting=args`....

The args is not really split but by setting type but passing it repeatedly and using duck typing style idea.

Reply:
See my latest commit for some code design idea. [try to resolve comments, may revert](https://github.com/ZidongLiu/DeComFL/pull/73/commits/5ed9088aa39cb39f06a7d1bfea544828a9c9084e)

I tried to make something like this, but I got some issue with the usage and unit test. we need to chat about it.

Reply:
resolved by adding @cache_property to all Settings